packages-ts/gauntlet-terra-contracts/src/lib/contracts.ts

```diff
@@ -28,109 +30,191 @@ export type TerraABI = {
-      return {
-        execute: io.readJSON(toPath('execute_msg')),
-        query: io.readJSON(toPath('query_msg')),
-        instantiate: io.readJSON(toPath('instantiate_msg')),
-      }
-    })
-  if (abi.length === 0) {
-    logger.error(`ABI not found for contract ${contractId}`)
+
+class Contracts {
+  // Retrieves a specific Contract object from the contract index, while loading its abi
+  // and bytecode from disk or network if they haven't been already.
+  async getContractWithSchemaAndCode(id: CONTRACT_LIST, version: string): Promise<Contract> {
+    const contract = this[id]
+    if (!contract) {
+      throw new Error(`Contract ${id} not found!`)
+    }
+    await Promise.all([
+      contract.abi ? Promise.resolve() : contract.loadContractABI(),
+      contract.bytecode ? Promise.resolve() : contract.loadContractCode(version),
+    ])
+    return contract
   }
 
-  return abi[0]
+  addChainlink = (id: CONTRACT_LIST, dirName: string) => {
+    this[id] = new ChainlinkContract(id, dirName)
+    return this
+  }
+
+  addCosmwasm = (id: CONTRACT_LIST, dirName: string) => {
+    this[id] = new CosmWasmContract(id, dirName)
+    return this
+  }
+}
+
+export const contracts = new Contracts()
+  .addChainlink(CONTRACT_LIST.FLAGS, 'flags')
+  .addChainlink(CONTRACT_LIST.DEVIATION_FLAGGING_VALIDATOR, 'deviation-flagging-validator')
+  .addChainlink(CONTRACT_LIST.OCR_2, 'ocr2')
+  .addChainlink(CONTRACT_LIST.PROXY_OCR_2, 'proxy-ocr2')
+  .addChainlink(CONTRACT_LIST.ACCESS_CONTROLLER, 'access-controller')
+  .addCosmwasm(CONTRACT_LIST.CW20_BASE, 'cw20_base')
+  .addCosmwasm(CONTRACT_LIST.CW4_GROUP, 'cw4_group')
+  .addCosmwasm(CONTRACT_LIST.MULTISIG, 'cw3_flex_multisig')
+
+type Instance = {
+  name: string
+  contractId: CONTRACT_LIST
 }
 
-export const getContract = async (id: CONTRACT_LIST, version): Promise<Contract> => {
-  version = version ? version : defaultContractVersions[id]
-  // Preload contracts
-  return {
-    id,
-    abi: getContractABI(id),
-    bytecode: await getContractCode(id, version),
+class AddressBook {
+  operator: AccAddress
+  instances: Map<AccAddress, Instance> // address => instance name
+
+  constructor() {
+    this.instances = new Map<AccAddress, Instance>()
+  }
+
+  addOperator(address: AccAddress) {
+    this.operator = address
+  }
+  addInstance(contractId: CONTRACT_LIST, envVar: string, name?: string) {
+    const address = process.env[envVar]
+    if (!address) {
+      console.warn(`${envVar} not set in environment--ignoring`)
+      return this // it's okay if we don't have all the contract addresses, for now
+    }
+
+    this.instances[address as AccAddress] = { name, contractId } as Instance
+
+    if (!AccAddress.validate(address))
+      throw new Error(`Read invalid contract address ${address} for ${contractId} contract from env`)
+
+    logger.debug(`Using deployed instance of ${contractId}: ${name}=${address}`)
+    return this
   }
 }
+
+export const addressBook = new AddressBook()
```

Comment:
Let's try to avoid global vars like this. 

The `addressBook` needs to be a member of a parent execution context. 

Perhaps a better place would be to actually add it to the `ExecutionContext` class, which should be ok for now, but that class is still too low of a level as it represents the context of executing one contract call, but the address book should be part of context representing the execution of Gauntlet binary as a whole (top-level).